oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/MongoRegexPathFilterFactory.java

```diff
@@ -0,0 +1,149 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.index.indexer.document.flatfile.pipelined;
+
+import org.apache.jackrabbit.oak.commons.PathUtils;
+import org.apache.jackrabbit.oak.spi.filter.PathFilter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.TreeSet;
+
+public class MongoRegexPathFilterFactory {
+
+    public static class MongoFilterPaths {
+        // Special value that means "download everything". This is used when regex path filtering is disabled or when
+        // the path filters would require downloading everything. When this is returned, it is preferable to do not
+        // use any filter in the Mongo query, even though using the values in this object as filters would also result
+        // in downloading everything
+        public static final MongoFilterPaths DOWNLOAD_ALL = new MongoFilterPaths(List.of("/"), List.of());
+
+        final List<String> included;
+        final List<String> excluded;
+
+        public MongoFilterPaths(List<String> included, List<String> excluded) {
+            this.included = included;
+            this.excluded = excluded;
+        }
+
+        @Override
+        public String toString() {
+            return "MongoFilterPaths{" +
+                    "included=" + included +
+                    ", excluded=" + excluded +
+                    '}';
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            MongoFilterPaths that = (MongoFilterPaths) o;
+            return Objects.equals(included, that.included) && Objects.equals(excluded, that.excluded);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(included, excluded);
+        }
+    }
+
+    private final static Logger LOG = LoggerFactory.getLogger(MongoRegexPathFilterFactory.class);
+
+    private final int maxNumberOfPaths;
+
+    public MongoRegexPathFilterFactory(int maxNumberOfPaths) {
+        this.maxNumberOfPaths = maxNumberOfPaths;
+    }
+
+    public MongoFilterPaths buildMongoFilter(List<PathFilter> pathFilters) {
```

Comment:
Question: From a quick look it seems that the computation done is largely duplicate of what is done in PathFilter(). Can't we just initialize a new PathFilter by concatenating the includes and excludes from the filters list?

Reply:
It is similar but not quite the same. For instance, if there is an index with 
```
IndexA
includedPaths=/a
excludedPaths=/a/a1

IndexB
includedPaths=/a, /b
excludedPaths=[]
```
Then `/a/a1` should not be excluded. But if we concatenate the included and excluded paths we get:
```
IndexPathsMerged
includedPaths=/a, /b
excludedPaths=/a/a1
```
which will exclude `/a/a1`.